### ISSUE TITLE
Can't load AppBridge v4 using NextJS app router. 

### ISSUE THREAD

**Issue by ozzyonfire** [Shopify/shopify-app-bridge#311]
## Describe the bug

Trying to load in the script tag for AppBridge v4. 
![image](https://github.com/Shopify/shopify-app-bridge/assets/779491/976385ff-6dc7-42b1-9e62-e841a4469b47)

The next/script component injects the script into the `<head>` before the page loads. Stategy = "beforeInteractive" ensures that everything is preloaded. 

Getting errors in the console about loading the script.
![image](https://github.com/Shopify/shopify-app-bridge/assets/779491/be65de3d-4916-4046-87bb-5d52e6a83e42)

However, the script does not have an async attribute.
![image](https://github.com/Shopify/shopify-app-bridge/assets/779491/a3438240-6838-4b26-bb9e-062fb566f879)

Although the script indeed does not appear first, I have no control over the order that the script is loaded (is this truly necessary?)

### To Reproduce

Create a next js starter app. Or checkout this branch of my repo. 
https://github.com/ozzyonfire/shopify-next-app/tree/app-bridge-v4

Steps to reproduce the behaviour:

Launch app and see error. 

### Expected behaviour

It should load app bridge. The script is being added to the head fine.

### Packages and versions

List the relevant packages you’re using, and their versions. For example:

- `@shopify/app-bridge` @ `4` ? script
- `@shopify/app-bridge-react` @ `4.0.0`
- `@shopify/polaris` @ `12.20.0`

## Platform

- OS: Macbook Pro - M2
- OS Version: 14.2.1
- App: Embedded App

### Additional context

Next.js v14.1.3

**Comment by khanakia** [comment]
I am having the same issue for now I just used raw head tag and it worked

```js
<head>
  {/* Next.js Script tag will not work cause of the async  and you wont
   * be able to use useAppBridge() hook otherwsie
   * Check this tutorial https://shopify.dev/docs/api/app-bridge/migration-guide
   */}
  <script src='https://cdn.shopify.com/shopifycloud/app-bridge.js'></script>
</head>
```

**Comment by khanakia** [comment]
Also, i cannot understand the difference between
1. `@shopify/app-bridge`
2. `https://cdn.shopify.com/shopifycloud/app-bridge.js`

Because without including the above cdn app-bridge.js file in the head below code does not work and give an error
```
Error: The Shopify global is not defined. This likely means the App Bridge script tag was not added correctly to this page
```

```js
import { Modal, TitleBar, useAppBridge } from '@shopify/app-bridge-react';
const shopify = useAppBridge();
```

if I included `@shopify/app-bridge` package already like this then why I need to use app bridge cdn script ?
```js
const app = createApp({
  apiKey: process.env.NEXT_PUBLIC_SHOPIFY_CLIENT_ID as string,
  host: host,
});
```

**Comment by TheSecurityDev** [comment]
You can just do a regular script tag and put it in the <Head> component. You will get warnings, but you can just ignore those.

**Comment by khanakia** [comment]
> I believe it is actually a next js issue. I did find a related bug that I expect to fix the problem.
> 
> [vercel/next.js#52935](https://github.com/vercel/next.js/issues/52935) - however this hasn't been patched yet from Next.js
> 
> Also, the @shopify/app-bridge is not v4 app bridge. There are two mechanisms on loading app bridge, CDN is the new way and the @shopify/app-bridge is the old way.

If `@shopify/app-bridge` is not v4 and I remove that package then how do I use the below code like subscribing to the event in my react.js app e.g.

```js
appState.getState().then((state: any) => {
 console.info('App State: %o', state);
});
    
app.subscribe(Redirect.Action.APP, (redirectData) => {
 console.log('aa', redirectData.path); // For example, '/settings'
});
```

**Comment by TheSecurityDev** [comment]
> > I believe it is actually a next js issue. I did find a related bug that I expect to fix the problem.
> > 
> > [vercel/next.js#52935](https://github.com/vercel/next.js/issues/52935) - however this hasn't been patched yet from Next.js
> > 
> > Also, the @shopify/app-bridge is not v4 app bridge. There are two mechanisms on loading app bridge, CDN is the new way and the @shopify/app-bridge is the old way.
> 
> If `@shopify/app-bridge` is not v4 and I remove that package then how do I use the below code like subscribing to the event in my react.js app e.g.
> 
> ```js
> appState.getState().then((state: any) => {
>  console.info('App State: %o', state);
> });
>     
> app.subscribe(Redirect.Action.APP, (redirectData) => {
>  console.log('aa', redirectData.path); // For example, '/settings'
> });
> ```

Is there a reason you need to do this? I found that it was no longer necessary.

But if you really want to, there's an undocumented feature you can use:
```js
shopify.protocol.subscribe("Navigation.redirect.app", (e) => console.log("Path: ", e.path))
```

**Comment by khanakia** [comment]
> I am having the same issue for now I just used raw head tag and it worked
> 
> ```js
> <head>
>   {/* Next.js Script tag will not work cause of the async  and you wont
>    * be able to use useAppBridge() hook otherwsie
>    * Check this tutorial https://shopify.dev/docs/api/app-bridge/migration-guide
>    */}
>   <script src='https://cdn.shopify.com/shopifycloud/app-bridge.js'></script>
> </head>
> ```

However, this approach worked in the development but now I cannot do `yarn build` next.js throws the error 

So any idea how to load the script wit next.js

**Comment by TheSecurityDev** [comment]
> > I am having the same issue for now I just used raw head tag and it worked
> > ```js
> > <head>
> >   {/* Next.js Script tag will not work cause of the async  and you wont
> >    * be able to use useAppBridge() hook otherwsie
> >    * Check this tutorial https://shopify.dev/docs/api/app-bridge/migration-guide
> >    */}
> >   <script src='https://cdn.shopify.com/shopifycloud/app-bridge.js'></script>
> > </head>
> > ```
> 
> However, this approach worked in the development but now I cannot do `yarn build` next.js throws the error
> 
> So any idea how to load the script wit next.js

What's the error you're getting?

**Comment by TheSecurityDev** [comment]
> @next/next/no-sync-scripts

I was getting this with eslint, so solved by adding an eslint ignore comment above:
```tsx
<Head>
  <meta name="shopify-api-key" content={process.env.SHOPIFY_API_KEY} />
  {/* eslint-disable-next-line @next/next/no-sync-scripts */}
  <script src="https://cdn.shopify.com/shopifycloud/app-bridge.js" />
</Head>
```

**Comment by khanakia** [comment]
Thanks build worked. Also, the app is working locally but when I deploy to Vercel and then load into Shopify Admin it gives this error

```js
The script tag loading App Bridge must be the first script tag in the document.
(anonymous) @ app-bridge.js:1
app-bridge.js:1 Uncaught Error: Shopify’s App Bridge must be included as the first <script> tag and must link to Shopify’s CDN. Do not use async, defer or type=module. Aborting.
    at app-bridge.js:1:38672
    at app-bridge.js:1:48313
    at app-bridge.js:1:55466
```

![Screenshot 2024-03-21 at 2 40 39 PM](https://github.com/Shopify/shopify-app-bridge/assets/2174170/a742ddc5-b793-4f9d-b711-f3f0b05b3159)

**Comment by TheSecurityDev** [comment]
No you have to add the attribute to the other script, if it's one you added. In this case it won't help you though.

What if you use the `<Head>` component from `next/link`? That's how I'm doing it.

**Comment by TheSecurityDev** [comment]
Do you have an `_app.jsx` file?

**Comment by TheSecurityDev** [comment]
> 1. I tried already head from next link i get this error
> 
> ![Screenshot 2024-03-21 at 9 33 36 PM](https://private-user-images.githubusercontent.com/2174170/315533071-f6fb1e4d-0cb4-4ed5-a85b-a5ea2a08e130.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEwMzc0MTIsIm5iZiI6MTcxMTAzNzExMiwicGF0aCI6Ii8yMTc0MTcwLzMxNTUzMzA3MS1mNmZiMWU0ZC0wY2I0LTRlZDUtYTg1Yi1hNWVhMmEwOGUxMzAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjFUMTYwNTEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGVmYmE2ZTdhNTFkZDA2MWVkNDg4MTdiMDI1Nzg4ZWIzZjRhMDFmODYxZDg5YjQ3MzljZDBjM2EwYmI3YWQyNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Ygz5cVMvkUhYTH9on-3KcoPjPpBNij836bLzQamnPgI)
> 
> 2. I am using app router so no i do not have `_app.tsx` file

I am only familiar with the pages router, so I don't really know. My Head component looks different on NextJS 14 being imported like `import Head from "next/head";`

**Comment by henrytao-me** [comment]
There are a lot of things could go wrong here. I try my best to summary what happens as below:

## TL;DR:
* You should use `<Head>` tag from `nextj/document` to add AB CDN script tag. See https://nextjs.org/docs/pages/building-your-application/routing/custom-document. 
* AB CDN requires to be added as the first script url in the page. 

> Also, i cannot understand the difference between `@shopify/app-bridge` and https://cdn.shopify.com/shopifycloud/app-bridge.js

As other explained above, you should just use AB from CDN and access all AB features without having to call `createApp` like the older version. 

> shopify.protocol.subscribe("Navigation.redirect.app", (data) => console.log("Path: ", data.path))

Please don't do this. It will be cleaned up in the future. If you want to subscribe to side nav click, you can use [ui-nav-menu](https://shopify.dev/docs/api/app-bridge-library/web-components/ui-nav-menu) and `addEventListener('click', () => {})` to each menu to handle click event.
